### DISCUSSION TITLE
Adding distance and/or level sensor class

### DISCUSSION THREAD

**Discussion by joakimeriksson** [home-assistant/architecture#607] [Entity Models] [answered]
## Context 
I am interested in using home assistant for measuring  the level of water in a drainage well. Either having a distance sensor class or a level sensor class would work. 

## Proposal
Add a sensor_class for distance and/or level to home assistant core and frontend (+ documentation).

## Consequence
I cut away from my previous PR (https://github.com/home-assistant/core/pull/54605) to only support distance and level and resubmit the PR (and add frontend icons and docs).

**Comment by MartinHjelmare** [comment]
I don't think distance is a good way to describe water level. Distance can use many different length unit of measurement depending on what kind of distance is measured. Just level is also too broad. I'd simply go with `"water_level"`.

Probably use meter and feet as unit of measurement.

**Reply by dgomes** [reply]
know of users that have diesel deposits for heating purposes, why not `liquid_level` ?

**Reply by MartinHjelmare** [reply]
If the same unit of measurement applies it could be ok. The risk with choosing a broader category is that the measurement may use a different unit. We don't want that.

**Reply by dgomes** [reply]
unit_of_measurement is usually Litres or M^3

@joakimeriksson knowing the volume of the well you can easily calculate the volume of water from your distance value

**Reply by joakimeriksson** [reply]
Yes, this is a very interesting discussion as distance is typically the actual "sensor" for getting the level and if you want litres or M^3 or in fact the quite common one: "percentage" (level / fullness 0 - 100). It kind of feels like we should have something else also - like "sensor class" + context => sensor = "liquid_level", context = drainage well or context = "diesel deposit".

But yes I would vote for liquid level - but I need distance to calculate it ;-)

**Reply by dgomes** [reply]
Not always distance, when it's a deposit an array of hall sensors usually does the trick.

**Reply by joakimeriksson** [reply]
Yes true - lots of options - weight or pressure would also be options.

**Comment by BradleyFord** [comment]
Remember that level cannot always be used for calculating volume or mass.
In many cases the sides of a tank are not vertical.

For example in an oil refinery the oil tanks do not have vertical sides,
they bulge due to the pressure from the contents.
Also we generally have a sump that is below the 0 point, and sediment at
the bottom. (Called BS&W)

So we always use a "strapping value", this is essentially a look up table
that shows the volume at key heights in the tank. Then we interpolate as
required.

Imagine a sphere or horizontal vessel, this becomes very important.


I spent about 5 years building oil moments systems in a few oil refineries,
getting this right was my core job.
We have special mechanical and radar guages that tell us oil level, water
level, sediment level. Then we can also estimate the oil in vapor, if it
has a fixed roof.

**Comment by joakimeriksson** [comment]
How do we conclude this one? Should I also look at any integration that would have any use of liquid_level?

**Reply by joakimeriksson** [reply]
Found some that could make use of liquid_level.

Ink sensor - homeassistant/components/epsonworkforce/sensor.py
Wemo fan - homeassistant/components/wemo/fan.py
Roomba - homeassistant/components/roomba/braava.py

There might be more but it seems like there is some existence of liquid level sensing already there in integrations.

**Reply by MartinHjelmare** [reply]
I think liquid level is too broad. Look at the examples above. An ink sensor in a printer won't have the same unit of measurement as a water level sensor in a well. We need to limit the measurement so that it represents the same type of measurements with the same unit of measurement.

**Reply by joakimeriksson** [reply]
I would say that using meters would be ok - for most types of sensors that would use an absolute measurement of liquid_level I guess meter would work - 0.1 meter in my well would be ok - while 1.2 meters would be really bad. Many liquid_levels would however be represented as 0-1 or 0-100% of "full". (car gas tanks, ink-cartridge, etc). For my well case any of the two would be ok - but we need to decide what the default is. Or is there a strict limitation on all liquid_levels having the same unit of measurement? And would then "percentage" be the best option? (quite generic and would work for most cases I would guess).

**Reply by dgomes** [reply]
car gas tank in % is not really good idea, existing integration already use Liters / Gallons

**Reply by joakimeriksson** [reply]
If we aim to be extremely specific we will end up with thousands of very very similar sensor classes - with same or similar code-base while if we go for very generic sensors (maybe level) they will not have anything in common. To find the right balance in this is very hard - is there any guiding document for this? Maybe we would need something guiding "rookies" like me here?!

**Reply by Jc2k** [reply]
I don't know of any guide, but one data point I have been using for device classes which may or may not be applicable here is what it means for integration with things like Siri or Alexa. For example, a volatile organic compounds device class is probably quite esoteric, and ordinarily I might not have wanted a dedicated device class and just set the unit of measurement. But by having it we can map sensors to HomeKit's AirQuality device type automatically, quite a nice usability win.

**Reply by joakimeriksson** [reply]
Ok, not a bad guiding principle. That will give support for most common types of sensors. I guess liquid level is not super-common - most sensor out there are rather flood sensor that level sensors when it comes to water - at least the ones commercially available on the consumer / smart home market. And flood or not is more of a binary sensor.

**Comment by BradleyFord** [comment]
I would recommend for now just doing liquid_level.

Then at some point in the future we can add the extra logic for handling
volume or mass with strapping tables.

**Comment by BradleyFord** [comment]
In my use case I am thinking about it more for my water tanks.

**Comment by joakimeriksson** [comment]
I guess we need to get a "final answer" here to do any progress and fix my old PR that included too much... I do not really see any complete convergence in the discussion - is there a timeout at some time - or will it be hanging until someone tags something as "the answer"?

**Reply by MartinHjelmare** [reply]
This suggestion is still too vague. We need to define the device class with a specific unit of measurement that is also useful for the measurement category we want to group. 

And please also list the integrations where this device class could be used.

Please update the top post when we have this information.

**Comment by epenet** [comment]
Would a `distance` device class be considered or is that still too broad?
Or maybe two device classes for `long_distance` (more than 1 meter) and `short_distance` (less than 10 meters)

Just looking at LENGTH_KILOMETERS, I can see that it would make sense for all the connected vehicle integrations out there (bmw, mazda, nissan, renault, starline, subaru...): Odometer/Mileage, Battery Autonomy/Range, Fuel Autonomy/Range

Another category using LENGTH_KILOMETERS is weather platforms for visibility (metoffice, environment_canada)

Could they be using the same device class?

**Comment by MartinHjelmare** [comment, marked as answer]
We've discussed this topic and concluded that it's ok to add a general sensor device class `distance` if we also implement custom unit selection option for the user and unit conversion, similar as we've done for the `pressure` sensor device class.

It's ok to go ahead with PRs that do this, for core, frontend, dev docs and user docs.

**Reply by epenet** [reply]
Opened https://github.com/home-assistant/core/pull/77951
